lib/elastic_record/percolator_model.rb

```diff
@@ -0,0 +1,44 @@
+module ElasticRecord
+  module PercolatorModel
+    def self.included(base)
+      base.class_eval do
+        class_attribute :percolates_model
+
+        include Model
+        extend ClassMethods
+      end
+    end
+
+    module ClassMethods
+      def elastic_index
+        @elastic_index ||=
+          begin
+            index = ElasticRecord::Index.new([self, percolates_model])
+            index.partial_updates = false
+            index
+          end
+      end
+
+      def doctype
+        @doctype ||= Doctype.percolator_doctype
+      end
+
+      def percolate(document)
+        query = {
+          "query" => {
+            "percolate" => {
+              "field"         => "query",
+              "document_type" => percolates_model.doctype.name,
+              "document"      => document
+            }
+          }
+        }
+
+        hits = elastic_index.search(query)['hits']['hits']
+        ids = hits.map { |hits| hits['_id'] }
+
+        where(id: ids)
+      end
```

Comment:
This is my only request: This method should take a model record, and not a hash.

The raw document stuff can live in `elastic_index`, or a separate method. My primary interest is that the public API supports a `Product` instance. (given that it implements `as_search_document`)

Reply:
I originally did this, but when it came to actually consuming the API, it made the percolation logic more complicated than necessary.  Presenting a document for percolation is not necessarily the responsibility of the class that represents the index query (PercolatorModel).

Reply:
Alright, I'll move on :)